doc/man/pvfs2-mkspace.1
.TH PVFS2-MKSPACE 1 2017-07-03
.SH NAME
\fBpvfs2-mkspace\fR \(en create storage space for OrangeFS server
.SH SYNOPSIS
\fBpvfs2-mkspace\fR [\fB\-adDhvV\fR] [\fB\-c \fIname\fR]
[\fB\-i \fIid\fR] [\fB\-r \fIhandle\fR] [\fB\-M \fIrange\fR]
[\fB\-N \fIrange\fR] [\fB\--data-space path\fR]
[\fB\--meta-space path\fR] \fIdirectory\fR
.SH DESCRIPTION
The
.B pvfs2-mkspace
utility creates an OrangeFS volume with a single collection or adds a
new collection to an existing storage space.  It should be used after
changing the storage configuration of a server which has already been
initialized.
.PP
The options are as follows:
.IP -a
Add a collection instead of creating a new storage space.
.IP -d
Use all default options.  The default options can be seen with
.BR -h .
.IP -D
Remove (permanently) the storage space.
.IP -h
Display synopsis.
.IP -v
Print version number and exit.
.IP -V
Display verbose messages.
.IP -c
Use
.I name
as the collection name.
.IP -i
Use
.I id
as file system ID.
.IP -r
Use
.I handle
as the root handle.
.IP -M
Use
.I range
as the metadata handle range.
.IP -N
Use
.I range
as the data handle range.
.IP --data-space
Use
.I path
as the path to the data space.
.IP --meta-space
Use
.I path
as the path to the metadata space.
.SH ENVIRONMENT
.IP PVFS2_DEBUGFILE
If set to the path of a local file, redirect debug output to it.
.IP PVFS2_DEBUGMASK
Set the OrangeFS debug mask.  Possible masks are documented in
.BR pvfs2-set-debugmask ( 1 ) \& .
.SH BUGS
Please submit bug reports to devel@lists.orangefs.org
.SH SEE ALSO
.BR pvfs2tab ( 5 )


ChangeLog
-----------------------

OrangeFS Release ChangeLog
-----------------------

orangefs-2.10.0 (Git Branch v.2.10.0)
https://github.com/waltligon/orangefs.git
up to commit-id 7c28362b

New functionality:

* - general updates
	1. numerous autotools related m4 files are updated.
	2. code refactored in numerous places to eliminate
	   compiler warnings from new versions of gcc.
	3. documentation updates.

* - Server
	1. updated formatting to conform to standard in simple-stripe code.

* - Clients
	1. build liborangefs by default. liborangefs provides an orangefs
	   version of posix system calls.

      2. Windows Client
         * Retargeted Client to the user mode file system library "dokany"
           v1.5.1.1000. (See https://github.com/dokan-dev/dokany)
         * Retargeted build to Visual Studio 2022.
         * Updated Event Logging for modern versions of Windows.
         * Removed obsolete security modes.
         * Updated the executable installer to install dokany and necessary
           Visual C++ runtime libraries.
         * Updated OpenSSL.
         * Made several quality-of-life improvements.

* - bmi
	1. Eliminated IBV_EVENT_WQ_FATAL warnings
	2. Added a new RDMA module:
		* Utilizes the RDMA Communication Manager library (librdmacm),
		  abstracting transport-specific details out of connection management
		* Adds RoCE v2 support to BMI
	3. Added a new experimental InfiniBand module for future development:
		* Still a work in progress and minimally tested
		* Current experimental changes include:
			- Improved connection-based reference counting




* - general fixes
	1. Fixed warning about using fast networks and TCP in pvfs2-genconfig
	

* - Client fixes
        1. Windows Client:
	     * Fixed a failure to delete files in some cases.
	     * Fixed compatibility issues between dokany 1.5.1 and the prior version.
	     * Fixed issues with error codes and directory-handling.




-----------------------
Development 2.9.9
orangefs-2.9.8 (Git Branch v.2.9.x)
https://github.com/waltligon/orangefs.git
up to rev 65ab0d2

This version of OrangeFS includes many bug fixes.

	1. Disable usrint by default.
	2. Fix potential memory overwrites in various utilities.
		- Client-core
		- Showcoll
		- Mkspace
		- Xattr
		- Viewdist
		- Str-utils
	3. Eliminate compiler warnings.
		- State machines
		- Trove
	4. Improve parsing of config file.
	5. Update documentation flow.
	6. Updated readmes.
	7. Migrate documentation to hugo format.
	8. Update perf mon example.



New functionality:
	1.  Added RoCE v1 support to our IB suite.


* - Server 
	1.  Fixed problems with crdirent and rename when using
	    distributed directories.
	2.  Fixed intermittent problems with updating timestamps.

* - Clients
	1.  pvfs2-stat.c
		When the object is a symlink, we now display the string length
		of the target in the size field, making the output consistent
		with the linux kernel.  Corrected call to sys-lookup so that
		symbolic links are followed correctly when -L is specified on
		the command line.
	2.  System Interface
		1.  Fixed problems with crdirent and rename when using distributed
		    directories.
		2.  Fixed intermittent problems with updating timestamps.

* - BMI
	1.  Part of adding RoCE v1 support was to make the IB port used for
	    communication configurable.

* - Software Infrastructure

* - Testing
	1.  Generic tests, #467 and #477, of the xfstest suite were removed from our
	    testing because we do not support the functionality.

	2.  Updated our testing to use version 20170929 of LTP.  Updated
	    the list of testcases to be used.

* - Kernel Module




orangefs-2.9.7 (Subversion Branch v.2.9.x)
http://dev.orangefs.org/svn/orangefs/branches/v.2.9.x
up to rev 13115 in trunk


New functionality:
	1.  Added support for Oracle Linux 7.3 with one caveat:  Cannot use
	    OrangeFS as the backing filesystem for a user-defined loop device.
	    Support will be forth coming.

	2.  Added support for large LMDB databases using DBMaxSize config
	    parameter.  

	3.  Added scripts in the examples/profile.d directory that can be placed in 
	    /etc/profile.d.  These scripts add OrangeFS locations to
	    LIBRARY_PATH, C_INCLUDE_PATH, AND LD_LIBRARY_PATH.

	4.  Added support for Fedora 26.  Certificate security is not yet
	    compatible with OpenSSL 1.1 but work is in progress.

	5.  Official OrangeFS rpm is in the Fedora distribution.  The rpm will
	    be in either Fedora 27 or Fedora 28.  The rpm will also be in
	    their EPEL repository, allowing other RHEL distributions to access
	    it.

	6.  Added --with-lmdb to the configure script, so you can specify the
	    location of your LMDB installation.  Otherwise, LMDB delivered
	    with OrangeFS will be used.  NOTE: Either way, to use LMDB instead
	    of Berkely DB, you must specify --with-db-backend=lmdb on the
	    configure line.

	7.  pvfs2-genconfig will ask you where you want to put the server config 
	    file, with the default being the <installation directory>/etc/orangefs.conf.
	    You can still put the server config filename directly on the command line.

	8.  ACL processing is now done as specified by POSIX.


* - Server  
	* - Enhanced permission checking and job scheduling for tree
	    operations.
	* - Removed memory leaks from the getattr operation and restructured
	    keyval buffer management.
	* - Corrected problem with mgmt-create-root-dir related to having a
	    correct capability when calling tree-setattr.
	* - Fixed capability errors that may occur during mkdir.
	* - Correctly populate the position token during a readdir operation
	    to prevent overrun of dirdata index bits.
	* - Correct the max size calculation for a readdir response.
	* - On some platforms, issuing a pvfs2-ls would cause one or more

src/apps/admin/pvfs2-mkspace.c
/*
 * (C) 2002 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <getopt.h>
#include <limits.h>
#include <string.h>
#include <stdlib.h>

#include "pvfs2.h"
#include "mkspace.h"
#include "pvfs2-internal.h"

#ifndef PVFS2_VERSION
#define PVFS2_VERSION "Unknown"
#endif

typedef struct
{
    int use_defaults;
    int verbose;
    PVFS_fs_id coll_id;
    PVFS_handle root_handle;
    int collection_only;
    int delete_storage;
    char collection[PATH_MAX];
    char data_space[PATH_MAX];
    char meta_space[PATH_MAX];
    char config_space[PATH_MAX];
} options_t;

static int default_verbose = 0;
static PVFS_handle default_root_handle = PVFS_HANDLE_NULL_INIT;
static int default_collection_only = 0;
static char default_collection[PATH_MAX] = "pvfs2-fs";

static void print_help(char *progname, options_t *opts);

static int parse_args(int argc, char **argv, options_t *opts)
{
    int ret = 0, option_index = 0;
    const char *cur_option = NULL;
    static struct option long_opts[] =
    {
        {"help",0,0,0},
        {"version",0,0,0},
        {"verbose",0,0,0},
        {"defaults",0,0,0},
        {"data-space",1,0,0},
        {"meta-space",1,0,0},
        {"config-space",1,0,0},
        {"coll-id",1,0,0},
        {"coll-name",1,0,0},
        {"root-handle",1,0,0},
        {"delete-storage",0,0,0},
        {"add-coll",0,0,0},
        {0,0,0,0}
    };

    if (argc == 1)
    {
        print_help(argv[0], opts);
        exit(1);
    }

    while ((ret = getopt_long(argc, argv, "c:i:r:vVhadD",
                              long_opts, &option_index)) != -1)
    {
	switch (ret)
        {
            case 0:
                cur_option = long_opts[option_index].name;

                if (strcmp("help", cur_option) == 0)
                {
                    goto do_help;
                }
                else if (strcmp("version", cur_option) == 0)
                {
                    goto do_version;
                }
                else if (strcmp("verbose", cur_option) == 0)
                {
                    goto do_verbose;
                }
                else if (strcmp("data-space", cur_option) == 0)
                {
                    strncpy(opts->data_space, optarg, PATH_MAX - 1);
                    break;
                }
                else if (strcmp("meta-space", cur_option) == 0)
		{
                    strncpy(opts->meta_space, optarg, PATH_MAX - 1);
                    break;
                }
                else if (strcmp("config-space", cur_option) == 0)
		{
                    strncpy(opts->config_space, optarg, PATH_MAX - 1);
                    break;
                }
                else if (strcmp("coll-id", cur_option) == 0)
                {
                    goto do_collection_id;
                }
                else if (strcmp("coll-name", cur_option) == 0)
                {
                    goto do_collection_name;
                }
                else if (strcmp("root-handle", cur_option) == 0)
                {
                    goto do_root_handle;
                }
                else if (strcmp("add-coll", cur_option) == 0)
                {
                    goto do_add_collection;
                }
                else if (strcmp("defaults", cur_option) == 0)
                {
                    goto do_defaults;
                }
                else if (strcmp("delete-storage", cur_option) == 0)
                {
                    goto do_delete_storage;
                }
                else
                {
                    print_help(argv[0], opts);
                    exit(1);
                }
	    case 'a':
          do_add_collection:
		opts->collection_only = 1;
		break;
	    case 'c':
          do_collection_name:
		strncpy(opts->collection, optarg, PATH_MAX - 1);
		break;
            case 'd':
          do_defaults:
                opts->use_defaults = 1;
                break;
            case 'h':
          do_help:
                print_help(argv[0], opts);
                exit(0);
	    case 'i':
          do_collection_id:
#ifdef HAVE_STRTOULL
		opts->coll_id = (PVFS_fs_id)strtoull(optarg, NULL, 10);
#else
		opts->coll_id = (PVFS_fs_id)strtoul(optarg, NULL, 10);
#endif
		break;
	    case 'r':
          do_root_handle:
                PVFS_OID_str2bin(optarg, &opts->root_handle);
		break;
	    case 'v':
          do_verbose:
		opts->verbose = PVFS2_MKSPACE_STDERR_VERBOSE;
		break;
            case 'V':
          do_version:
                fprintf(stderr,"%s\n",PVFS2_VERSION);
                exit(0);
            case 'D':
          do_delete_storage:
                opts->delete_storage = 1;
                break;
	    case '?':
                fprintf(stderr, "%s: error: unrecognized "
                        "option '%c'.\n", argv[0], ret);
	    default:
                print_help(argv[0], opts);
		return -1;
	}
    }
    return 0;
}

static void print_options(options_t *opts)
{
    if (opts)
    {
        printf("\t   use all defaults    : %s\n",
               (opts->use_defaults ? "yes" : "no"));
        printf("\t   delete storage      : %s\n",
               (opts->delete_storage ? "yes" : "no"));
        printf("\t   verbose             : %s\n",
               (opts->verbose ? "ON" : "OFF"));
        printf("\t   root handle         : %s\n",
               PVFS_OID_str(&opts->root_handle));
        printf("\t   collection-only mode: %s\n",
               (opts->collection_only ? "ON" : "OFF"));
        printf("\t   collection id       : %d\n", opts->coll_id);
        printf("\t   collection name     : %s\n",
               (strlen(opts->collection) ?
                opts->collection : "None specified"));
        printf("\t   data storage space  : %s\n",
               (strlen(opts->data_space) ?
                opts->data_space : "None specified"));
        printf("\tmetadata storage space : %s\n",
	       	(strlen(opts->meta_space) ?
		opts->meta_space : "None specified"));
        printf("\tconfig storage space : %s\n",
	       	(strlen(opts->config_space) ?
		opts->config_space : "None specified"));
    }
}

static void print_help(char *progname, options_t *opts)
{
    fprintf(stderr,"usage: %s [OPTION]...\n", progname);
    fprintf(stderr,"This program is useful for creating a new pvfs2 "
            "storage space with a \nsingle collection, or adding a "
            "new collection to an existing storage space.\n\n");
    fprintf(stderr,"The following arguments can be used to "
           "customize your volume:\n");

    fprintf(stderr,"  -a, --add-coll                       "
            "used to add a collection only\n");
    fprintf(stderr,"  -c, --coll-name                      "
            "create collection with the speciifed name\n");
    fprintf(stderr,"  -d, --defaults                       "
            "use all default options (see below)\n");
    fprintf(stderr,"  -D, --delete-storage                 "
            "REMOVE the storage space (unrecoverable!)\n");
    fprintf(stderr,"  -h, --help                           "
            "show this help listing\n");
    fprintf(stderr,"  -i, --coll-id=ID                     "
            "create collection with the specified id\n");
    fprintf(stderr,"  -r, --root-handle=HANDLE             "
            "create collection with this root handle\n");
    fprintf(stderr,"       --data-space=PATH             "
            "create data storage space at this location\n");
    fprintf(stderr,"       --meta-space=PATH             "
            "create metadata storage space at this location\n");
    fprintf(stderr,"       --config-space=PATH             "
            "create config storage space at this location\n");
    fprintf(stderr,"  -v, --verbose                        "
            "operate in verbose mode\n");
    fprintf(stderr,"  -V, --version                        "
            "print version information and exit\n");

    fprintf(stderr,"\nIf the -d or --defaults option is used, the "
            "following options will be used:\n");
    opts->use_defaults = 1;
    print_options(opts);
}

int main(int argc, char **argv)
{
    int ret = -1;
    options_t opts;
    memset(&opts, 0, sizeof(options_t));

    if (parse_args(argc, argv, &opts))
    {
	fprintf(stderr,"%s: error: argument parsing failed; "
                "aborting!\n", argv[0]);
	return -1;
    }

    if (opts.use_defaults)
    {
        opts.verbose = default_verbose;
        opts.root_handle = default_root_handle;
        opts.collection_only = default_collection_only;
        strncpy(opts.collection, default_collection, PATH_MAX);
    }

    print_options(&opts);

    if (strlen(opts.data_space) == 0)
    {
        fprintf(stderr, "Error: You MUST specify a data storage space\n");
        return -1;
    }

    if (strlen(opts.meta_space) == 0)
    {
        fprintf(stderr, "Error: You MUST specify a metadata storage space\n");
        return -1;
    }

    if (strlen(opts.config_space) == 0)
    {
        fprintf(stderr, "Error: You MUST specify a config storage space\n");
        return -1;
    }

    if (opts.coll_id == PVFS_FS_ID_NULL)
    {
        fprintf(stderr, "Error: You MUST specify a collection ID\n");
        return -1;
    }

    if (opts.delete_storage)
    {
        ret = pvfs2_rmspace(opts.data_space,
                            opts.meta_space,
                            opts.config_space,
			    opts.collection,
                            opts.coll_id, 
			    opts.collection_only,
                            opts.verbose);
    }
    else
    {
        printf("opts.collection_only(%d).\n", opts.collection_only);
        ret = pvfs2_mkspace(opts.data_space,
                            opts.meta_space,
                            opts.config_space,
			    opts.collection,
                            opts.coll_id, 
                            PVFS_HANDLE_NULL,
                            PVFS_HANDLE_NULL,
                            NULL,
                            0,
                            opts.collection_only, 
			    opts.verbose);
    }
    return ret;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
